README.md
# Attendance

[![Build Status](https://img.shields.io/travis/aminebenhariz/Attendance/master.svg?style=flat)](https://travis-ci.org/aminebenhariz/Attendance)
[![Code Coverage](https://scrutinizer-ci.com/g/aminebenhariz/Attendance/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/aminebenhariz/Attendance/?branch=master)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/aminebenhariz/Attendance/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/aminebenhariz/Attendance/?branch=master)

[![SensioLabsInsight](https://insight.sensiolabs.com/projects/432625e3-2487-4f67-91fa-25d6fd54db67/big.png)](https://insight.sensiolabs.com/projects/432625e3-2487-4f67-91fa-25d6fd54db67)

## Table of Contents

+ [Install](#install)
+ [Usage](#usage)
+ [DayAttendanceLine Format](#dayattendanceline-format)
+ [Requirements](#requirements)
+ [Testing](#testing)

## Install

Via Composer

``` bash
$ composer require aminebenhariz/attendance
```

## DayAttendanceLine Format

```
2015-12-12|08:30 (10:00-10:30) (16:00-16:30) 17:30

 - 2015-12-12    : date (yyyy-mm-dd)
 - 08:30         : arrival time (hh:mm)
 - (10:00-10:30) : pause from 10:00 to 10:30
 - (16:00-16:30) : pause from 16:00 to 16:30
 - 17:30         : departure time (hh:mm)
```

Description can be added like this

```
2015-12-12|08:30 (10:00-10:30) (16:00-16:30) 17:30|Some Description
```

## Usage

### Calculate attendance

``` php
$dayAttendanceLine = '2015-12-12|08:30 (10:00-10:30) (16:00-16:30) 17:30';
$dayAttendance = DayAttendance::parseDayAttendanceLine($dayAttendanceLine);

echo $dayAttendance->getDuration()->format('%H:%I:%S');
// 07:30:00

echo $dayAttendance->exportLine();
// 2015-12-12|08:30 (10:00-10:15) (12:30-13:30) (16:00-16:15) 17:30
```

### Calculate average attendance of multiple days

``` php
$day1 = DayAttendance::parseDayAttendanceLine('2015-12-14|08:31 (12:02-13:42) 17:25');
$day2 = DayAttendance::parseDayAttendanceLine('2015-12-15|08:29 (12:21-13:32) (16:12-16:22) 17:24');
$day3 = DayAttendance::parseDayAttendanceLine('2015-12-16|08:52 (12:18-13:12) 17:31');
$day4 = DayAttendance::parseDayAttendanceLine('2015-12-17|08:12 (12:21-13:52) 17:24');
$day5 = DayAttendance::parseDayAttendanceLine('2015-12-18|08:35 (12:24-13:25) 17:42');

$attendance = new Attendance([$day1, $day2, $day3, $day4, $day5]);

echo $attendance->getAverage()->format('%H:%I');
// 07:40
```

## Requirements

The following versions of PHP are supported by this version.

+ PHP 5.4
+ PHP 5.5
+ PHP 5.6
+ PHP 7.0
+ HHVM

## Testing

``` bash
$ composer test
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.4.1] - 2016-03-29
### Fixed
- Reduce code complexity

## [0.4.0] - 2016-01-19
### Added
- Add optional description to DayAttendance

## [0.3.0] - 2015-12-17
### Added
- `AmineBenHariz\Attendance\Attendance`
- Calculate average attendance (multiple days)

## [0.2.0] - 2015-12-16
### Added
- The changelog you are reading.
- Parse DayAttendanceLine.
- Export DayAttendanceLine.

## 0.1.0 - 2015-12-16
### Added
- `AmineBenHariz\Attendance\Pause`
- `AmineBenHariz\Attendance\DayAttendance`

[Unreleased]: https://github.com/aminebenhariz/Attendance/compare/0.4.1...HEAD
[0.4.1]: https://github.com/aminebenhariz/Attendance/compare/0.4.0...0.4.1
[0.4.0]: https://github.com/aminebenhariz/Attendance/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/aminebenhariz/Attendance/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/aminebenhariz/Attendance/compare/0.1.0...0.2.0

tests/PauseTest.php
<?php

namespace AmineBenHariz\Attendance\Tests;

use AmineBenHariz\Attendance\Pause;

/**
 * Class PauseTest
 * @package AmineBenHariz\Attendance\Tests
 */
class PauseTest extends \PHPUnit_Framework_TestCase
{
    /**
     * @dataProvider invalidPauseCreationProvider
     * @param \DateTime $start
     * @param \DateTime $end
     * @expectedException \InvalidArgumentException
     */
    public function testInvalidPauseCreation(\DateTime $start, \DateTime $end)
    {
        new Pause($start, $end);
    }

    /**
     * @return array
     */
    public function invalidPauseCreationProvider()
    {
        return [
            // Pause ends before even starting! (may be possible in quantum mechanics)
            [new \DateTime('2015-12-12 10:00'), new \DateTime('2015-12-12 09:30')],

            // Pause ends the next day. Are you that lazy?
            [new \DateTime('2015-12-12 10:00'), new \DateTime('2015-12-13 10:30')],
        ];
    }

    /**
     * @return Pause
     */
    public function testValidPauseCreation()
    {
        $start = new \DateTime('2015-12-12 09:30');
        $end = new \DateTime('2015-12-12 10:00');

        $pause = new Pause($start, $end);

        $this->assertSame($start, $pause->getStart());
        $this->assertSame($end, $pause->getEnd());

        return $pause;
    }

    /**
     * @dataProvider pauseOverlappingProvider
     * @param Pause $pause1
     * @param Pause $pause2
     * @param $overlapping
     */
    public function testPauseOverlapping(Pause $pause1, Pause $pause2, $overlapping)
    {
        // isOverlapping is symmetric, must work both ways
        $this->assertSame($overlapping, $pause1->isOverlapping($pause2));
        $this->assertSame($overlapping, $pause2->isOverlapping($pause1));
    }

    /**
     * @return array
     */
    public function pauseOverlappingProvider()
    {
        return [
            [
                // no overlapping
                //   ======
                //             =======
                new Pause(new \DateTime('2015-12-12 09:30'), new \DateTime('2015-12-12 10:00')),
                new Pause(new \DateTime('2015-12-12 10:30'), new \DateTime('2015-12-12 11:00')),
                false,
            ],
            [
                // intersection
                //   ======
                //        =======
                new Pause(new \DateTime('2015-12-12 09:30'), new \DateTime('2015-12-12 10:00')),
                new Pause(new \DateTime('2015-12-12 09:45'), new \DateTime('2015-12-12 10:15')),
                true,
            ],
            [
                // contained
                //       ======
                //     ===========
                new Pause(new \DateTime('2015-12-12 09:30'), new \DateTime('2015-12-12 11:00')),
                new Pause(new \DateTime('2015-12-12 10:00'), new \DateTime('2015-12-12 10:30')),
                true,
            ],
            [
                // just touching, this is not considered overlapping
                //    ======
                //          =======
                new Pause(new \DateTime('2015-12-12 09:30'), new \DateTime('2015-12-12 10:00')),
                new Pause(new \DateTime('2015-12-12 10:00'), new \DateTime('2015-12-12 10:30')),
                false,
            ],
        ];
    }

    /**
     * @depends testValidPauseCreation
     * @param Pause $pause
     */
    public function testGetPauseDuration(Pause $pause)
    {
        $this->assertSame('00:30:00', $pause->getDuration()->format('%H:%I:%S'));
    }

    /**
     * @depends testValidPauseCreation
     * @param Pause $pause
     */
    public function testExportBlock(Pause $pause)
    {
        $this->assertSame('(09:30-10:00)', $pause->exportBlock());
    }
}


tests/AttendanceTest.php
<?php

namespace AmineBenHariz\Attendance\Tests;

use AmineBenHariz\Attendance\Attendance;
use AmineBenHariz\Attendance\DayAttendance;

/**
 * Class AttendanceTest
 * @package AmineBenHariz\Attendance\Tests
 */
class AttendanceTest extends \PHPUnit_Framework_TestCase
{

    /**
     * @expectedException \InvalidArgumentException
     */
    public function testInvalidAttendanceCreation()
    {
        // 2 lines with same date
        $dayAttendance1 = DayAttendance::parseDayAttendanceLine('2015-12-14|08:30 (12:30-13:30) 17:30');
        $dayAttendance2 = DayAttendance::parseDayAttendanceLine('2015-12-15|08:30 (12:30-13:30) 17:30');
        $dayAttendance3 = DayAttendance::parseDayAttendanceLine('2015-12-16|08:30 (12:30-13:30) 17:30');
        $dayAttendance4 = DayAttendance::parseDayAttendanceLine('2015-12-17|08:30 (12:30-13:30) 17:30');
        $dayAttendance5 = DayAttendance::parseDayAttendanceLine('2015-12-17|08:30 (12:30-13:30) 17:30');

        $dayAttendanceList = [
            $dayAttendance1,
            $dayAttendance2,
            $dayAttendance3,
            $dayAttendance4,
            $dayAttendance5,
        ];

        new Attendance($dayAttendanceList);
    }

    /**
     * @return Attendance
     */
    public function testValidAttendanceCreation()
    {
        $dayAttendance1 = DayAttendance::parseDayAttendanceLine('2015-12-14|08:31 (12:02-13:42) 17:25');
        $dayAttendance2 = DayAttendance::parseDayAttendanceLine('2015-12-15|08:29 (12:21-13:32) (16:12-16:22) 17:24');
        $dayAttendance3 = DayAttendance::parseDayAttendanceLine('2015-12-16|08:52 (12:18-13:12) 17:31');
        $dayAttendance4 = DayAttendance::parseDayAttendanceLine('2015-12-17|08:12 (12:21-13:52) 17:24');
        $dayAttendance5 = DayAttendance::parseDayAttendanceLine('2015-12-18|08:35 (12:24-13:25) 17:42');

        $dayAttendanceList = [
            $dayAttendance1,
            $dayAttendance2,
            $dayAttendance3,
            $dayAttendance4,
            $dayAttendance5,
        ];

        $attendance = new Attendance($dayAttendanceList);

        $this->assertInstanceOf('AmineBenHariz\Attendance\Attendance', $attendance);
        $this->assertCount(5, $attendance->getDayAttendanceList());

        return $attendance;
    }

    /**
     * @depends testValidAttendanceCreation
     * @param Attendance $attendance
     */
    public function testGetTotalMinutes(Attendance $attendance)
    {
        $this->assertSame(2300, $attendance->getTotalMinutes());
    }

    /**
     * @depends testValidAttendanceCreation
     * @depends testGetTotalMinutes
     * @param Attendance $attendance
     */
    public function testGetAverage(Attendance $attendance)
    {
        $this->assertSame('07:40', $attendance->getAverage()->format('%H:%I'));

        // test empty attendance
        $emptyAttendance = new Attendance();
        $this->assertSame('00:00', $emptyAttendance->getAverage()->format('%H:%I'));
    }
}


tests/DayAttendanceTest.php
<?php

namespace AmineBenHariz\Attendance\Tests;

use AmineBenHariz\Attendance\DayAttendance;
use AmineBenHariz\Attendance\Pause;

/**
 * Class DayAttendanceTest
 * @package AmineBenHariz\Attendance\Tests
 */
class DayAttendanceTest extends \PHPUnit_Framework_TestCase
{
    /**
     * @dataProvider invalidDayAttendanceCreationProvider
     * @param \DateTime $arrival
     * @param \DateTime $departure
     * @param Pause[] $pauseList
     * @expectedException \InvalidArgumentException
     */
    public function testInvalidDayAttendanceCreation(\DateTime $arrival, \DateTime $departure, array $pauseList)
    {
        new DayAttendance($arrival, $departure, $pauseList);
    }

    /**
     * @return array
     */
    public function invalidDayAttendanceCreationProvider()
    {
        return [
            [
                // arrival and departure not in the same day
                new \DateTime('2015-12-12 08:30'),
                new \DateTime('2015-12-13 17:30'),
                []
            ],
            [
                // arrival after departure
                new \DateTime('2015-12-12 17:30'),
                new \DateTime('2015-12-12 08:30'),
                []
            ],
            [
                // pause starts before arrival
                new \DateTime('2015-12-12 08:30'),
                new \DateTime('2015-12-12 17:30'),
                [
                    new Pause(new \DateTime('2015-12-12 07:00'), new \DateTime('2015-12-12 07:30')),
                    new Pause(new \DateTime('2015-12-12 10:00'), new \DateTime('2015-12-12 10:30')),
                ]
            ],
            [
                // pause ends after departure
                new \DateTime('2015-12-12 08:30'),
                new \DateTime('2015-12-12 17:30'),
                [
                    new Pause(new \DateTime('2015-12-12 10:00'), new \DateTime('2015-12-12 10:30')),
                    new Pause(new \DateTime('2015-12-12 17:15'), new \DateTime('2015-12-12 17:45')),
                ]
            ],
            [
                // overlapping pauses
                new \DateTime('2015-12-12 08:30'),
                new \DateTime('2015-12-12 17:30'),
                [
                    new Pause(new \DateTime('2015-12-12 10:00'), new \DateTime('2015-12-12 10:30')),
                    new Pause(new \DateTime('2015-12-12 10:15'), new \DateTime('2015-12-12 10:45')),
                ]
            ],
        ];
    }

    /**
     * @return DayAttendance
     */
    public function testValidDayAttendanceCreation()
    {
        $arrival = new \DateTime('2015-12-12 08:30');
        $departure = new \DateTime('2015-12-12 17:30');

        $pause1 = new Pause(new \DateTime('2015-12-12 10:00'), new \DateTime('2015-12-12 10:15'));
        $pause2 = new Pause(new \DateTime('2015-12-12 12:30'), new \DateTime('2015-12-12 13:30'));
        $pause3 = new Pause(new \DateTime('2015-12-12 16:00'), new \DateTime('2015-12-12 16:15'));
        $pauseList = [$pause1, $pause2, $pause3];

        $dayAttendance = new DayAttendance($arrival, $departure, $pauseList);
        $dayAttendance->setDescription('Some description');

        $this->assertSame($arrival, $dayAttendance->getArrival());
        $this->assertSame($departure, $dayAttendance->getDeparture());
        $this->assertSame($pauseList, $dayAttendance->getPauseList());
        $this->assertSame('Some description', $dayAttendance->getDescription());

        return $dayAttendance;
    }

    /**
     * @depends testValidDayAttendanceCreation
     * @param DayAttendance $dayAttendance
     */
    public function testGetDuration(DayAttendance $dayAttendance)
    {
        $this->assertSame('07:30:00', $dayAttendance->getDuration()->format('%H:%I:%S'));
    }

    /**
     * @dataProvider isValidDayAttendanceLineProvider
     * @param string $dayAttendanceLine
     * @param bool $valid
     */
    public function testIsValidDayAttendanceLine($dayAttendanceLine, $valid)
    {
        $this->assertSame($valid, DayAttendance::isValidDayAttendanceLine($dayAttendanceLine));
    }

    /**
     * @return array
     */
    public function isValidDayAttendanceLineProvider()
    {
        return [
            ['', false], // empty
            ['lorem', false], // random
            ['08:30 (10:00-10:30) (16:00-16:30) 17:30', false], // no date
            ['2015-12-12|', false], // no timeLine
            ['2015-12-12|(10:00-10:30) (16:00-16:30) 17:30', false], // no arrival
            ['2015-12-12|08:30 (10:00-10:30) (16:00-16:30)', false], // no departure
            ['2015-12-1x|08:30 (10:00-10:30) (16:00-16:30) 17:30', false], // invalid date

            ['2015-12-12|08:30 (10:00-10:30) (16:00-16:30) 17:30', true], // correct
            ['2015-12-12|08:30 17:30', true], // correct, no pauses
        ];
    }

    /**
     * @return DayAttendance
     */
    public function testParseDayAttendanceLineWithDescription()
    {
        $dayAttendanceLine = '2015-12-12|08:30 (10:00-10:30) (16:00-16:30) 17:30|Some description';
        $dayAttendance = DayAttendance::parseDayAttendanceLine($dayAttendanceLine);

        $this->assertInstanceOf('\AmineBenHariz\Attendance\DayAttendance', $dayAttendance);

        $this->assertSame('2015-12-12 08:30', $dayAttendance->getArrival()->format('Y-m-d H:i'));
        $this->assertSame('2015-12-12 17:30', $dayAttendance->getDeparture()->format('Y-m-d H:i'));

        $pauseList = $dayAttendance->getPauseList();
        $this->assertCount(2, $pauseList);

        $this->assertSame('2015-12-12 10:00', $pauseList[0]->getStart()->format('Y-m-d H:i'));
        $this->assertSame('2015-12-12 10:30', $pauseList[0]->getEnd()->format('Y-m-d H:i'));

        $this->assertSame('2015-12-12 16:00', $pauseList[1]->getStart()->format('Y-m-d H:i'));
        $this->assertSame('2015-12-12 16:30', $pauseList[1]->getEnd()->format('Y-m-d H:i'));

        $this->assertSame('Some description', $dayAttendance->getDescription());

        return $dayAttendance;
    }

    /**
     * @return DayAttendance
     */
    public function testParseDayAttendanceLineWithoutDescription()
    {
        $dayAttendanceLine = '2015-12-12|08:30 (10:00-10:30) (16:00-16:30) 17:30';
        $dayAttendance = DayAttendance::parseDayAttendanceLine($dayAttendanceLine);

        $this->assertInstanceOf('\AmineBenHariz\Attendance\DayAttendance', $dayAttendance);

        $this->assertSame('2015-12-12 08:30', $dayAttendance->getArrival()->format('Y-m-d H:i'));
        $this->assertSame('2015-12-12 17:30', $dayAttendance->getDeparture()->format('Y-m-d H:i'));

        $pauseList = $dayAttendance->getPauseList();
        $this->assertCount(2, $pauseList);

        $this->assertSame('2015-12-12 10:00', $pauseList[0]->getStart()->format('Y-m-d H:i'));
        $this->assertSame('2015-12-12 10:30', $pauseList[0]->getEnd()->format('Y-m-d H:i'));

        $this->assertSame('2015-12-12 16:00', $pauseList[1]->getStart()->format('Y-m-d H:i'));
        $this->assertSame('2015-12-12 16:30', $pauseList[1]->getEnd()->format('Y-m-d H:i'));

        $this->assertSame('', $dayAttendance->getDescription());

        return $dayAttendance;
    }

    /**
     * @return DayAttendance
     */
    public function testParseDayAttendanceLineWithoutPauses()
    {
        $dayAttendanceLine = '2015-12-12|08:30 17:30';
        $dayAttendance = DayAttendance::parseDayAttendanceLine($dayAttendanceLine);

        $this->assertInstanceOf('\AmineBenHariz\Attendance\DayAttendance', $dayAttendance);

        $this->assertSame('2015-12-12 08:30', $dayAttendance->getArrival()->format('Y-m-d H:i'));
        $this->assertSame('2015-12-12 17:30', $dayAttendance->getDeparture()->format('Y-m-d H:i'));

        $pauseList = $dayAttendance->getPauseList();
        $this->assertCount(0, $pauseList);

        $this->assertSame('', $dayAttendance->getDescription());

        return $dayAttendance;
    }

    /**
     * @expectedException \InvalidArgumentException
     */
    public function testInvalidParseDayAttendanceLine()
    {
        $dayAttendanceLine = 'lorem';
        DayAttendance::parseDayAttendanceLine($dayAttendanceLine);
    }

    /**
     * @depends testParseDayAttendanceLineWithDescription
     * @param DayAttendance $dayAttendance
     */
    public function testExportLineWithDescription(DayAttendance $dayAttendance)
    {
        $this->assertSame(
            '2015-12-12|08:30 (10:00-10:30) (16:00-16:30) 17:30|Some description',
            $dayAttendance->exportLine()
        );
    }

    /**
     * @depends testParseDayAttendanceLineWithoutDescription
     * @param DayAttendance $dayAttendance
     */
    public function testExportLineWithoutDescription(DayAttendance $dayAttendance)
    {
        $this->assertSame(
            '2015-12-12|08:30 (10:00-10:30) (16:00-16:30) 17:30',
            $dayAttendance->exportLine()
        );
    }

    /**
     * @depends testValidDayAttendanceCreation
     * @param DayAttendance $dayAttendance
     */
    public function testGetDate(DayAttendance $dayAttendance)
    {
        $this->assertSame('2015-12-12', $dayAttendance->getDate());
    }

    /**
     * @depends testValidDayAttendanceCreation
     * @param DayAttendance $dayAttendance
     */
    public function testGetTimeLine(DayAttendance $dayAttendance)
    {
        $this->assertSame(
            '08:30 (10:00-10:15) (12:30-13:30) (16:00-16:15) 17:30',
            $dayAttendance->getTimeLine()
        );
    }
}


src/Pause.php
<?php

namespace AmineBenHariz\Attendance;

/**
 * Class Pause
 * @package AmineBenHariz\Attendance
 */
class Pause
{
    /**
     * @var \DateTime
     */
    private $start;

    /**
     * @var \DateTime
     */
    private $end;

    /**
     * Pause constructor.
     * @param \DateTime $start
     * @param \DateTime $end
     */
    public function __construct(\DateTime $start, \DateTime $end)
    {
        if ($start > $end) {
            throw new \InvalidArgumentException("Pause can't start after ending");
        }

        if ($start->format('Y-m-d') !== $end->format('Y-m-d')) {
            throw new \InvalidArgumentException("Pause must end in the same day");
        }

        $this->start = $start;
        $this->end = $end;
    }

    /**
     * @return \DateTime
     */
    public function getStart()
    {
        return $this->start;
    }

    /**
     * @return \DateTime
     */
    public function getEnd()
    {
        return $this->end;
    }

    /**
     * @return \DateInterval
     */
    public function getDuration()
    {
        return $this->getStart()->diff($this->getEnd());
    }

    /**
     * @param Pause $pause
     * @return bool
     */
    public function isOverlapping(Pause $pause)
    {
        if ($this->getEnd() <= $pause->getStart()) {
            return false;
        }

        if ($this->getStart() >= $pause->getEnd()) {
            return false;
        }

        return true;
    }

    /**
     * @return string
     */
    public function exportBlock()
    {
        return '(' . $this->getStart()->format('H:i') . '-' . $this->getEnd()->format('H:i') . ')';
    }
}


src/Attendance.php
<?php

namespace AmineBenHariz\Attendance;

/**
 * Class Attendance
 * @package AmineBenHariz\Attendance
 */
class Attendance
{
    /**
     * @var DayAttendance[]
     */
    private $dayAttendanceList = [];

    /**
     * Attendance constructor.
     * @param DayAttendance[] $dayAttendanceList
     */
    public function __construct(array $dayAttendanceList = [])
    {
        foreach ($dayAttendanceList as $dayAttendance) {
            $this->addDayAttendance($dayAttendance);
        }
    }

    /**
     * @return DayAttendance[]
     */
    public function getDayAttendanceList()
    {
        return $this->dayAttendanceList;
    }

    /**
     * @param DayAttendance $dayAttendance
     */
    private function addDayAttendance(DayAttendance $dayAttendance)
    {
        $date = $dayAttendance->getArrival()->format('Y-m-d');

        if (isset($this->dayAttendanceList[$date])) {
            throw new \InvalidArgumentException;
        }

        $this->dayAttendanceList[$date] = $dayAttendance;
    }

    /**
     * @return int
     */
    public function getTotalMinutes()
    {
        $totalMinutes = 0;

        foreach ($this->getDayAttendanceList() as $dayAttendance) {
            $totalMinutes += $dayAttendance->getTotalMinutes();
        }

        return $totalMinutes;
    }

    /**
     * @return \DateInterval
     */
    public function getAverage()
    {
        $getAttendanceList = $this->getDayAttendanceList();
        if (empty($getAttendanceList)) {
            return new \DateInterval('PT0S');
        }

        $totalMinutes = $this->getTotalMinutes();

        $averageMinutes = $totalMinutes / count($getAttendanceList);

        return new \DateInterval('PT' . floor($averageMinutes / 60) . 'H' . $averageMinutes % 60 . 'M');
    }
}


src/DayAttendance.php
<?php

namespace AmineBenHariz\Attendance;

/**
 * Class DayAttendance
 * @package AmineBenHariz\Attendance
 */
class DayAttendance
{
    /**
     * example: 2015-12-12|08:30 (10:00-10:30) (12:30-13:30) (16:00-16:30) 17:30
     */
    const DAY_ATTENDANCE_LINE_REGEX =
        '/^\d{4}-\d{2}-\d{2}\|\d{2}:\d{2}( \(\d{2}:\d{2}-\d{2}:\d{2}\))* \d{2}:\d{2}(\|[^\|]*)?$/';

    /**
     * @var \DateTime
     */
    private $arrival;

    /**
     * @var \DateTime
     */
    private $departure;

    /**
     * @var Pause[]
     */
    private $pauseList = [];

    /**
     * @var string
     */
    private $description = "";

    /**
     * DayAttendance constructor.
     * @param \DateTime $arrival
     * @param \DateTime $departure
     * @param Pause[] $pauseList
     */
    public function __construct(\DateTime $arrival, \DateTime $departure, array $pauseList = [])
    {
        if ($arrival > $departure || $arrival->format('Y-m-d') !== $departure->format('Y-m-d')) {
            throw new \InvalidArgumentException;
        }

        $this->arrival = $arrival;
        $this->departure = $departure;

        foreach ($pauseList as $pause) {
            $this->addPause($pause);
        }
    }

    /**
     * @return \DateTime
     */
    public function getArrival()
    {
        return $this->arrival;
    }

    /**
     * @return \DateTime
     */
    public function getDeparture()
    {
        return $this->departure;
    }

    /**
     * @return Pause[]
     */
    public function getPauseList()
    {
        return $this->pauseList;
    }

    /**
     * @return string
     */
    public function getDescription()
    {
        return $this->description;
    }

    /**
     * @param string $description
     */
    public function setDescription($description)
    {
        $this->description = $description;
    }

    /**
     * @param Pause $pause
     */
    private function addPause(Pause $pause)
    {
        if ($pause->getStart() < $this->getArrival()) {
            throw new \InvalidArgumentException;
        }

        if ($pause->getEnd() > $this->getDeparture()) {
            throw new \InvalidArgumentException;
        }

        if ($this->isPauseOverlapping($pause)) {
            throw new \InvalidArgumentException;
        }


        $this->pauseList[] = $pause;
    }

    /**
     * @param Pause $pause
     * @return bool
     */
    private function isPauseOverlapping(Pause $pause)
    {
        $existingPauseList = $this->getPauseList();

        foreach ($existingPauseList as $existingPause) {
            if ($pause->isOverlapping($existingPause)) {
                return true;
            }
        }

        return false;
    }

    /**
     * @return \DateInterval
     */
    public function getDuration()
    {
        $cursor = clone $this->getArrival();

        // PHP 5.4 : empty() can only handle variables
        $pauseList = $this->getPauseList();

        foreach ($pauseList as $pause) {
            $cursor->add($pause->getDuration());
        }

        return $cursor->diff($this->getDeparture());
    }

    /**
     * @return int
     */
    public function getTotalMinutes()
    {
        $duration = $this->getDuration();
        return intval($duration->format('%H')) * 60 + intval($duration->format('%I'));
    }

    /**
     * @param $dayAttendanceLine
     * @return DayAttendance
     */
    public static function parseDayAttendanceLine($dayAttendanceLine)
    {
        if (!self::isValidDayAttendanceLine($dayAttendanceLine)) {
            throw new \InvalidArgumentException;
        }

        list($date, $timeLine, $description) = self::explodeDayAttendanceLineParts($dayAttendanceLine);

        $times = explode(' ', $timeLine);

        $arrival = new \DateTime($date . ' ' . array_shift($times));
        $departure = new \DateTime($date . ' ' . array_pop($times));

        $pauseList = self::parsePauseBlocks($date, $times);

        $dayAttendance = new DayAttendance($arrival, $departure, $pauseList);
        $dayAttendance->setDescription($description);
        return $dayAttendance;
    }

    /**
     * @param string $dayAttendanceLine
     * @return array
     */
    private static function explodeDayAttendanceLineParts($dayAttendanceLine)
    {
        $parts = explode('|', $dayAttendanceLine);

        // if empty description
        if (!isset($parts[2])) {
            $parts[2] = '';
        }

        return $parts;
    }

    /**
     * @param string $date date in Y-m-d format, ex: 2016-03-29
     * @param array $pauseBlocks
     * @return Pause[]
     */
    private static function parsePauseBlocks($date, $pauseBlocks)
    {
        $pauseList = [];
        foreach ($pauseBlocks as $pauseBlock) {
            $pauseList[] = self::parsePauseBlock($date, $pauseBlock);
        }

        return $pauseList;
    }

    /**
     * @param string $date date in Y-m-d format, ex: 2016-03-29
     * @param string $pauseBlock ex: (10:00-10:30)
     * @return Pause
     */
    private static function parsePauseBlock($date, $pauseBlock)
    {
        $pauseStart = new \DateTime($date . ' ' . substr($pauseBlock, 1, 5));
        $pauseEnd = new \DateTime($date . ' ' . substr($pauseBlock, 7, 5));
        return new Pause($pauseStart, $pauseEnd);
    }

    /**
     * @param $dayAttendanceLine
     * @return int
     */
    public static function isValidDayAttendanceLine($dayAttendanceLine)
    {
        return preg_match(self::DAY_ATTENDANCE_LINE_REGEX, $dayAttendanceLine) === 1;
    }

    /**
     * @return string
     */
    public function exportLine()
    {
        $line = $this->getDate() . '|' . $this->getTimeLine();

        if ($this->getDescription() !== '') {
            $line .= '|' . $this->getDescription();
        }

        return $line;
    }

    /**
     * @return string
     */
    public function getDate()
    {
        return $this->getArrival()->format('Y-m-d');
    }

    /**
     * @return string
     */
    public function getTimeLine()
    {
        $line = $this->getArrival()->format('H:i');

        foreach ($this->getPauseList() as $pause) {
            $line .= ' ' . $pause->exportBlock();
        }

        $line .= ' ' . $this->getDeparture()->format('H:i');

        return $line;
    }
}
